include/opengemini/impl/enc/LineProtocolEncoder.cpp

```diff
@@ -0,0 +1,185 @@
+    }
+
+    AppendEscapeString(measurement, ESCAPE_CHARS_MEASUREMENT);
+}
+
+OPENGEMINI_INLINE_SPECIFIER
+void LineProtocolEncoder::AppendTags(const decltype(Point::tags)& tags)
+{
+    for (auto& [key, value] : tags) {
+        Append(ELEMENT_COMMA);
+        AppendEscapeString(key, ESCAPE_CHARS_TAG);
+        Append(ELEMENT_EQUAL);
+        AppendEscapeString(value, ESCAPE_CHARS_TAG);
+    }
+}
+
+OPENGEMINI_INLINE_SPECIFIER
+void LineProtocolEncoder::AppendFields(const decltype(Point::fields)& fields)
+{
+    if (fields.empty()) {
+        throw Exception(errc::LogicErrors::InvalidArgument,
+                        "The filed <fields> in Point must not be empty");
+    }
+
+    Append(ELEMENT_SPACE);
+    std::for_each_n(fields.begin(),
+                    fields.size() - 1,
+                    [this](const auto& field) {
+                        AppendField(field);
+                        Append(ELEMENT_COMMA);
+                    });
+    AppendField(*fields.rbegin());
+}
+
+OPENGEMINI_INLINE_SPECIFIER
+void LineProtocolEncoder::AppendTimestamp(const Point::Time& time,
+                                          Precision          precision)
+{
+    using namespace std::chrono;
+
+    int64_t count{ 0 };
+    switch (precision) {
+    case Precision::Nanosecond: count = time.time_since_epoch().count(); break;
+    case Precision::Microsecond: count = Round<microseconds>(time); break;
+    case Precision::Millisecond: count = Round<milliseconds>(time); break;
+    case Precision::Second: count = Round<seconds>(time); break;
+    case Precision::Minute: count = Round<minutes>(time); break;
+    case Precision::Hour: count = Round<hours>(time); break;
+    }
+
+    if (count != 0) {
+        Append(ELEMENT_SPACE);
+        Append(count);
+    }
+}
+
+OPENGEMINI_INLINE_SPECIFIER
+void LineProtocolEncoder::AppendField(
+    const decltype(Point::fields)::value_type& field)
+{
+    auto& [key, value] = field;
+
+    AppendEscapeString(key, ESCAPE_CHARS_FIELD_KEY);
+    Append(ELEMENT_EQUAL);
+    std::visit(
+        [this](const auto& innerValue) {
+            using T = std::decay_t<decltype(innerValue)>;
+
+            if constexpr (std::is_same_v<T, std::string>) {
+                Append(ELEMENT_DQUOTE);
+                AppendEscapeString(innerValue, ESCAPE_CHARS_FIELD_VALUE);
+                Append(ELEMENT_DQUOTE);
+            }
+            else if constexpr (std::is_same_v<T, uint64_t>) {
+                Append(innerValue);
+                Append(ELEMENT_UINT);
+            }
+            else if constexpr (std::is_same_v<T, int64_t>) {
+                Append(innerValue);
+                Append(ELEMENT_INT);
+            }
+            else if constexpr (std::is_same_v<T, bool>) {
+                Append(innerValue ? ELEMENT_TRUE : ELEMENT_FALSE);
+            }
+            else if constexpr (std::is_same_v<T, double>) {
+                Append(innerValue);
+            }
```

Comment:
shoud we add a default error? like

```c++
else {
    static_assert(always_false<T>::value, "Unhandled type in std::variant.");
}
```

Reply:
Good advice, I will add a runtime checking in the else branch.